File: lib/services/ingredient_factory.dart
import '../data/food_knowledge.dart';
import '../models/ingredient.dart';
import '../models/shopping_item.dart';
import '../models/storage_area.dart';

abstract final class IngredientFactory {
  static Ingredient fromShoppingItem(ShoppingItem item, {DateTime? now}) {
    final defaults = FoodKnowledge.lookup(item.name);
    final addedAt = now ?? DateTime.now();
    final shelfLifeDays = defaults?.shelfLifeDays;
    final expiryDate =
        shelfLifeDays == null
            ? null
            : addedAt.add(Duration(days: shelfLifeDays));

    return Ingredient(
      name: item.name,
      quantity: '1',
      unit: '份',
      imageUrl: item.imageUrl ?? '',
      freshnessPercent: expiryDate == null ? 0.85 : 1.0,
      state: FreshnessState.fresh,
      category: FoodKnowledge.categoryFor(item.name),
      storage: defaults?.storage ?? IconType.fridge,
      expiryDate: expiryDate,
      addedAt: addedAt,
      shelfLifeDays: shelfLifeDays,
      expiryLabel: expiryDate == null ? '新鲜' : '$shelfLifeDays天后过期',
    );
  }
}

File: apps/ios/FreshPantry/Domain/Proposals/IngredientFactory.swift
import Foundation

/// Constructs an inventory `Ingredient` from a `ShoppingItem`, applying
/// `FoodKnowledge` defaults (category/storage/shelf-life). Ported VERBATIM from
/// `lib/services/ingredient_factory.dart`.
///
/// `id` is left empty (local rows have empty id per the identity invariant).
enum IngredientFactory {
    static func fromShoppingItem(_ item: ShoppingItem, now: Date = Date()) -> Ingredient {
        let defaults = FoodKnowledge.lookup(item.name)
        let addedAt = now
        let shelfLifeDays = defaults?.shelfLifeDays
        let expiryDate = shelfLifeDays.map { addedAt.addingTimeInterval(TimeInterval($0 * 86400)) }

        return Ingredient(
            name: item.name,
            quantity: "1",
            unit: "份", // i18n:ignore data identity, not UI text
            imageUrl: item.imageUrl ?? "",
            freshnessPercent: expiryDate == nil ? 0.85 : 1.0,
            state: .fresh,
            expiryLabel: expiryDate == nil ? String(localized: "expiry.fresh") : String(localized: "expiry.inDays \(shelfLifeDays!)"),
            category: FoodKnowledge.categoryFor(item.name),
            storage: defaults?.storage ?? .fridge,
            expiryDate: expiryDate,
            addedAt: addedAt,
            shelfLifeDays: shelfLifeDays
        )
    }
}
